### DISCUSSION TITLE
Vitest 1.0 tests hang (no errors) after finishing tests, nothing from hanging-process reporter

### DISCUSSION THREAD

**Discussion by knownasilya** [vitest-dev/vitest#4797] [Q&A] [answered]
I have vitest setup but it hangs and doesn't close, I added 'hanging-process' reporter but get nothing from that

```ts
// vite.config.mts
// ...
test: {
    environment: 'jsdom',
    include: ['**/*.test.ts', '**/*.test.tsx'],
    setupFiles: [path.resolve(__dirname, 'vitest.setup.ts')],
    // fixes issue with testing-library/react not cleaning up after each run
    globals: true,
    reporters: ['default', 'hanging-process'],
  },
```

Anything else I can do? I tried `--pool forks` that still hangs 
Also I couldn't use `--pool forks --poolOptions.forks.isolate false` because that broke rendering tests with testing-library/react.

(this is a migration in progress from jest) 

```json
"vite": "^5.0.10",
"vite-plugin-mkcert": "^1.17.1",
"vite-plugin-node-polyfills": "^0.17.0",
"vitest": "^1.1.0",
"@vitejs/plugin-react": "^4.2.1",
```

How to debug this and find the source of problem? Jest had `detectOpenHandles` but I don't see anything similar except the hanging-process reporter which doesn't even print anything. The process just hangs forever doesn't even timeout.

**Comment by AriPerkkio** [comment, marked as answer]
> How to debug this and find the source of problem?

- Try removing additional Vite plugins
- Remove any `vi.mock` calls: https://github.com/vitest-dev/vitest/issues/4373
- Check source files for circular imports
- Update Node to latest
- Reduce test case into a minimal one, remove all possible sensitive info and post a minimal reproduction case here. By doing this you'll be helping others as well.

**Reply by knownasilya** [reply]
I had one `importOriginal` and removed it. I rewrote it to something like:

```ts
vi.mock('tg/tg-api');

vi.mocked(api.updateTask).mockImplementation(() => Promise.resolve({} as Task));
vi.mocked(api.getProjectChildren).mockImplementation(() => Promise.resolve([]));
```

Looks like I have a few failing tests still. Thanks for the help!

**Comment by linonetwo** [comment]
`close timed out after 10000ms / FILEHANDLE (unknown stack trace)`

This happens because Electron/Vitest child processes do not inherit the ELECTRON_RUN_AS_NODE environment variable, so resources cannot be cleaned up and handles leak.

Do not set `ELECTRON_RUN_AS_NODE` in `vitest.config.ts` via `process.env.ELECTRON_RUN_AS_NODE = 'true'` — this only affects the main process, not child processes.

Always use cross-env in your test script. For example:

`cross-env ELECTRON_RUN_AS_NODE=1 pnpm exec ./node_modules/.bin/electron ./node_modules/vitest/vitest.mjs run`

Or run manually in shell: `$env:ELECTRON_RUN_AS_NODE=1; pnpm run test:unit`

We use `ELECTRON_RUN_AS_NODE` to solve native modules (like better-sqlite3) being compiled for the wrong Node.js version https://github.com/WiseLibs/better-sqlite3/issues/545#issuecomment-809710688
